src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,351 @@
+                let c2 = cmp_ctx.clone();
+                let c3 = mul_ctx.clone();
+                let record_id = RecordId::from(i);
+
+                async move {
+                    let delta_bits = BitDecomposition::execute(c1, record_id, rbg, delta).await?;
+                    let compare_bit = one
+                        - &bitwise_greater_than_constant(c2, record_id, &delta_bits, cap.into())
+                            .await?;
+                    c3.multiply(record_id, &row.trigger_value, &compare_bit)
+                        .await
+                }
+            }),
+    )
+    .await
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Step {
+    IsTriggerBitTimesHelperBit,
+    InitializeTimeDelta,
+    PreviousStopBitTimesCurrent,
+    RandomBitsForBitDecomposition,
+    TimeDeltaBitDecomposition,
+    TimeDeltaLessThanCap,
+    CompareBitTimesTriggerValue,
+}
+
+impl crate::protocol::Substep for Step {}
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::IsTriggerBitTimesHelperBit => "is_trigger_bit_times_helper_bit",
+            Self::InitializeTimeDelta => "initialize_time_delta",
+            Self::PreviousStopBitTimesCurrent => "previous_stop_bit_times_current",
+            Self::RandomBitsForBitDecomposition => "random_bits_for_bit_decomposition",
+            Self::TimeDeltaBitDecomposition => "time_delta_bit_decomposition",
+            Self::TimeDeltaLessThanCap => "time_delta_less_than_cap",
+            Self::CompareBitTimesTriggerValue => "compare_bit_times_trigger_value",
+        }
+    }
+}
+
+#[cfg(all(test, not(feature = "shuttle")))]
+mod tests {
+    use crate::attribution_window_test_input;
+    use crate::ff::{Field, Fp32BitPrime};
+    use crate::protocol::attribution::{
+        apply_attribution_window::apply_attribution_window,
+        input::{
+            ApplyAttributionWindowInputRow, MCApplyAttributionWindowInputRow,
+            MCApplyAttributionWindowOutputRow,
+        },
+    };
+    use crate::protocol::context::Context;
+    use crate::protocol::modulus_conversion::{convert_all_bits, convert_all_bits_local};
+    use crate::protocol::{BreakdownKey, MatchKey};
+    use crate::secret_sharing::replicated::semi_honest::AdditiveShare as Replicated;
+    use crate::secret_sharing::SharedValue;
+    use crate::test_fixture::input::GenericReportTestInput;
+    use crate::test_fixture::{Reconstruct, Runner, TestWorld};
+
+    #[tokio::test]
+    pub async fn attribution_window() {
+        const ATTRIBUTION_WINDOW: u32 = 600;
+        const EXPECTED: &[u128; 23] = &[
+            0, 0, 0, 10, 2, 1, 5, 1, 0, 0, 0, 10, 0, 3, 12, 0, 0, 6, 4, 0, 6, 1, 0,
+        ];
+
+        let input: Vec<GenericReportTestInput<Fp32BitPrime, MatchKey, BreakdownKey>> = attribution_window_test_input!(
+            [
+                { timestamp: 500, is_trigger_report: 0, helper_bit: 0, breakdown_key: 3, credit: 0 }, // delta: 0
```

Comment:
Are these actually timestamps, or are they deltas?

If they are actually timestamps, then 500 should not come before 100. That was super confusing to me.

Reply:
Oh I see... it's from a different user... so it's OK. wow this is subtle.